### ISSUE TITLE
Type inference problem with first and last

### ISSUE THREAD

**Issue by cartant** [ReactiveX/rxjs#3946]
## Bug Report

**Current Behavior**

The changes in PR #3945 exposed a bug with typings for the `first` and `last` operators.

With those operators, it's possible to specify a default. And if one is specified, its type is used to infer both the operator's input and output types. And, in doing so, the inferred `MonoTypeOperatorFunction` type can be incompatible with the source observable.

**Reproduction**

See [this test](https://github.com/ReactiveX/rxjs/blob/6.2.2/spec/operators/first-spec.ts#L135) for `first`:

```ts
expectObservable(e1.pipe(first(x => x === 's', 'd'))).toBe(expected);
```

**Expected behavior**

Because the default value is specified as using a string literal, the inferred type is a literal type: `MonoTypeOperatorFunction<"d">` and that's not compatible with the source observable's type of `Observable<string>`.

**Possible Solution**

The current workaround used for the test is to explicitly specify the type parameter:

```ts
expectObservable(e1.pipe(first<string>(x => x === 's', 'd'))).toBe(expected);
```

Instead of using `T` as the type for the `default` parameter, it ought to be possible to use another type parameter that extends `T`. Something like this, maybe:

```ts
export function first<T, D extends T = T>(
  predicate?: null,
  defaultValue?: D
): MonoTypeOperatorFunction<T>;
/* ... etc. ... */
```

**Comment by cartant** [comment]
FWIW, here's a minimal repro of the problem:

```ts
interface Foo<T> {}
interface Bar<T, S> {}

function foo<T>(predicate?: null, defaultValue?: T): Foo<T>;
function foo<T, S extends T>(predicate: (value: T) => value is S, defaultValue?: T): Bar<T, S>;
function foo<T>(predicate: (value: T) => boolean, defaultValue?: T): Foo<T>;
function foo<T>(predicate?: ((value: T) => boolean) | null, defaultValue?: T): Foo<T> | Bar<T, T> {
  return undefined!;
}
const result = foo(x => x === 's', 'd'); // Foo<"d">
```

If inserted into the RxJS project - say, after the implementation of `first` - `result` is inferred as `Foo<"d">`.

However, I've been unable to reproduce this in another project, so it's possible that it's a bug that's only effected with a particular combination of TypeScript version and `compilerOptions`. Or something.

**Comment by cartant** [comment]
Opened a TypeScript issue for the problem: https://github.com/Microsoft/TypeScript/issues/25917

**Comment by cartant** [comment]
The weird, compiler-options-dependent behaviour appears to have been fixed in TypeScript 3.0.
